#47 Removed presentation role from TimeInput

emilyuhde opened the pull request:
Since children of the div do have semantic meaning, this role was causing an accessibility violation.

Fix can be verified using the HTML CodeSniffer browser bookmark I mention in a comment on issue #47 

Closes #47

emilyuhde commented:
@wojtekmaj I'm having some issues with the build failing for a linter rule. When I ran the unit tests locally with `npm run test`, I was getting errors that state needed to be initialized in constructors. I changed them to be in constructors to resolve the unit test warnings, but now the build is failing saying that state should be initialized in class properties, not in constructors. I just want to check with you, am I doing something wrong here? I can change the code back to having the state initialized as class properties, but then is there something that needs to be updated with the unit tests so they no longer give errors?

wojtekmaj commented on sample/parcel/Sample.jsx:
```diff
@@ -4,8 +4,11 @@ import TimePicker from 'react-time-picker';
 export default class Sample extends Component {
-  state = {
-    value: new Date(),
+  constructor(props) {
```
Why the change in constructors?

emilyuhde replied:
Ok, thanks! I'll change the state initialization back and update the PR.

wojtekmaj commented:
Hmmm... I look at this implementation and maybe all we need is to remove `role="presentation"`, that's it? Not even adding the new role? onKeyDown doesn't make a lot of sense in this context, and even may cause input focus issues 🤔

emilyuhde commented:
@wojtekmaj Sure, if there does not need to be a role on the div, I can remote it and that will make things simpiler. There was an accessibility error happening when I changed the role to group saying that if there was an onClick handler there needed to be some keyboard input handler as well, which is why I added the onKeyDown. With no role, the onKeyDown handler is no longer needed.

I made the changes (removing the role and onKeyDown) and leaving the onClick handler where I moved it to and everything seems to be working the same as before and there are no accessibility errors reported.

wojtekmaj commented:
Sadly it doesn't look like everything works as expected :( This breaks 26275b8c033e0b9a4b8edfa59a36fb443c65351c which was the purpose of this onClick in the first place.

Looks like the only way would be to leave the code exactly as it is, remove the `role`, and disable ESLint rules responsible for the errors:


```jsx

    /* eslint-disable jsx-a11y/click-events-have-key-events */
    /* eslint-disable jsx-a11y/no-static-element-interactions */
    return (
      <div
        className={className}
        onClick={this.onClick}
      >
        {this.renderNativeInput()}
        {this.renderCustomInputs()}
      </div>
    );
```

What do you think?

emilyuhde commented:
That's too bad it breaks that commit :-/ 

I think your suggestion is good. I think the role="presentation" might cause actual accessibility problems with screen readers, but the static elements being interactive and key events errors don't seem to actually be reporting anything being broken for screen reader users. You can keyboard nav to all of the TimePicker elements just fine.

Let me try making that change you suggested and make sure there aren't any other strange side effects of just removing the role.

wojtekmaj commented:
I am aware of the dangers of "div onClick" but I think in this case it should be very much okay. This is just a helper for people NOT using accessibility helpers, who might on an empty space in Time Picker. Expected action, as in native input, is to select the first field ([try that!](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/input/time)). People using assistive technologies by definition can't click on an empty space by accident, so this action is of no use to them.

wojtekmaj commented:
Congrats on your contribution!

emilyuhde commented on sample/parcel/package.json:
```diff
@@ -16,7 +16,7 @@
     "prop-types": "^15.6.2",
     "react": "^16.5.0",
     "react-dom": "^16.5.0",
-    "react-time-picker": "file:react-time-picker-3.7.2.tgz"
```
Sorry, I thought I had excluded this change from my commits!

wojtekmaj replied:
No worries. Next time you can use Yarn, and play with yarn link command - much easier way to test stuff :)
